kotlinx-coroutines-core/common/src/channels/Channels.common.kt

```diff
@@ -49,10 +49,44 @@ public fun <E : Any> ReceiveChannel<E>.onReceiveOrNull(): SelectClause1<E?> {
 }
 
 /**
- * Makes sure that the given [block] consumes all elements from the given channel
- * by always invoking [cancel][ReceiveChannel.cancel] after the execution of the block.
+ * Executes the [block] and then [cancels][ReceiveChannel.cancel] the channel,
+ * ensuring that all the elements that were sent are processed by either [block] or [ReceiveChannel.cancel].
  *
- * The operation is _terminal_.
+ * It is guaranteed that, after invoking this operation, the channel will be [cancelled][ReceiveChannel.cancel], so
+ * the operation is _terminal_.
+ * If the [block] finishes with an exception, that exception will be used for cancelling the channel.
+ *
+ * This function is useful for building more complex terminal operators while ensuring that no elements will be lost.
+ * Example:
+ *
+ * ```
+ * suspend fun <E> ReceiveChannel<E>.consumeFirst(): E =
+ *     consume { return receive() }
+ * fun Int.cleanup() { println("cleaning up $this") }
+ * val channel = Channel<Int>(10, onUndeliveredElement = Int::cleanup)
+ * // Launch a procedure that creates values
+ * launch(Dispatchers.Default) {
+ *     repeat(10) {
+ *         val sendResult = channel.trySend(it)
+ *         if (sendResult.isFailure) {
+ *             print("in the producer: ")
+ *             it.cleanup()
+ *         }
+ *         yield()
+ *     }
+ * }
+ * // Grab the first value and discard everything else
+ * launch(Dispatchers.Default) {
+ *     val firstElement = channel.consumeFirst()
+ *     println("received $firstElement")
+ * }
+ * ```
+ *
+ * In this example, all ten values created by the producer coroutine will be processed: one by `consumeFirst`,
```

Comment:
Reading this, it seems like the example is mixing two concepts: one is about consuming, and the other about undelivered clean up (which we don't consider as "processed", it has a different meaning).

Maybe something more straightforward?
Like the following (outline, haven't tried to compile it):
```
val producer = produce {
    var currentElement = 0
    try {
        while(true) {
            send(currentElement++)
        }
    } finally {
        println("Cancelled after: $currentElement")
    }
}

consumeFirst { ... }
```

It's deterministic, more trivial and straightforward. WDYT?

Reply:
Yes, good point. The point of `consume` seems to be just notifying the producer that there are no more consumers. I noticed that `consume` is a reliable way not to miss any elements and became too fixated on that, but this really is a just side effect.